076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,389 @@
+
+> Fig 6: The numbers in the diagram match the numbers in the full description of the steps.
+> The diagram omits the use of annotations and the EntityOperator, CruiseControl and KafkaExporter reconcilers, and the removal of trust of the old CA certificate and key.
+
+Notes:
+* The existing annotations strimzi.io/ca-cert-generation and strimzi.io/server-cert-hash are still used to keep track of the certificate being presented by the Kubernetes Pod and which generation of the Cluster CA signed that certificate.
+* The strimzi.io/ca-key-generation is no longer needed, because it is being tracked in the trust-state Kubernetes Secret.
+
+### Future changes
+
+This proposal does not cover all changes needed to integrate with certificate management solutions.
+Instead it aims to make an improvement to the existing logic and act as a first step towards integrating some certificate management solution.
+The approach suggested aligns with the approach taken in the [closed proposal on CA abstraction][pr46].
+
+Other/Future proposals will be needed to cover:
+ - Refactoring the end-entity certificate issuance process, so that components request a certificate and asynchronously are issued a certificate.
+ - Refactoring the tracking of which certificates are signed by which CA certificates.
+   Currently, the notion of a "generation" is used, however if Strimzi is no longer managing the CA this concept will need to be rethought.
+ - Refactoring the CA certificate management process to allow alternative implementations internally, for example using the cert-manager API.
+
+## Affected/not affected projects
+
+This affects the strimzi-kafka-operator project only.
+
+## Compatibility
+
+When the operator is upgraded to a version that includes this proposal it will automatically use this new mechanism for all new and existing Kafka clusters.
+This means for existing Kafka clusters once the operator is upgraded it will automatically generate new certificates for the components.
+This is because the operator has no way of knowing whether the trust-state Kubernetes Secret is missing because this is a new cluster, or because we are doing an upgrade.
+This will only happen the first time upgrading the operator, since in future the trust-state Kubernetes Secret will already exist.
+
+For an existing Kafka cluster the flow will be (this is the same whether the CA is managed by Strimzi or by the user):
+
+1. The CaReconciler takes the Cluster CA certificate from the <CLUSTER_NAME>-cluster-ca-cert Kubernetes Secret and places it in the trust-state Kubernetes Secret with the state UNTRUSTED.
+2. The CaReconciler places the Cluster CA certificate in the trusted-certs Kubernetes Secret.
+3. The CaReconciler rolls the Kafka and ZooKeeper Kubernetes Pods since it believes they do not trust the new CA certificate, then it updates the trust-state Kubernetes Secret to put the CA certificate into TRUSTED_UNUSED.
```

Comment:
This is definitely not something to do on the upgrade as that will likely screw many things up by how non-restartable that rolling update is. Also, I'm not sure they should start with this state. Maybe they should start with TRUSTED_IN_USE?